airbyte-integrations/connectors/source-harvest/source_harvest/streams.py

```diff
@@ -0,0 +1,361 @@
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional
+
+import pendulum
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+
+
+class HarvestStream(HttpStream, ABC):
+    url_base = "https://api.harvestapp.com/v2/"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        pass
+
+
+class HarvestStreamWithPagination(HarvestStream):
+    per_page = 50
+    primary_key = "id"
+
+    @property
+    @abstractmethod
+    def data_field(self) -> str:
+        """
+        :return: Default field name to get data from response
+        """
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        stream_data = response.json()
+        if stream_data["next_page"] is not None:
+            return {
+                "page": stream_data["next_page"],
+            }
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        params["per_page"] = self.per_page
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        records = response.json()[self.data_field]
+        for record in records:
+            yield record
+
+
+class HarvestStreamWithPaginationSliced(HarvestStreamWithPagination):
+    @property
+    @abstractmethod
+    def parent_stream(self) -> HarvestStreamWithPagination:
+        """
+        :return: parent stream class
+        """
+
+    def stream_slices(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, any]]]:
+        updated_since = self._updated_since
+        if stream_state and stream_state.get(self.cursor_field):
+            updated_since = stream_state[self.cursor_field]
+        items = self.parent_stream(authenticator=self.authenticator, updated_since=updated_since)
+        for item in items.read_records(sync_mode=sync_mode):
```

Comment:
It seems to me that this is not true.
If I understand correctly, then we take the state of the child stream and get the parent stream to start recording using this state, and then the child.
However, I'm sure that child streams can change and be added regardless of the parent stream.
For example, if we have already received data from the "`invoice_payments`" stream once, and start the reading process again, then we will not receive a single entry for the parent stream (`invoice`), because the parent stream is always created before the child stream. And in this case, even if new `invoice_payments` were added, but the parent stream was not updated, we will not receive new records.

Please ping me if I expressed myself not quite clearly - I will explain more clearly)